### DISCUSSION TITLE
adCoupledGradient in FV interface kernel

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#22590] [Q&A General] [answered]
I am working on an FV interface kernel that needs to compute the difference in an (elemental) adCoupledGradient. That is, I have an FV elemental variable  (an adCoupledValue in the interface kernel) and an adCoupledGradient that is the gradient of that FV elemental variable, and I need to evaluate the adCoupledGradient on both sides of the interface. I have not quite been able to figure out how to do that. For adFunctors I can just evaluate them at elemFromFace and neighborFromFace. If I assume that I can pass components of the gradient as variable1 and variable2, the I can use the FVInterfaceKernel system to generate differences in the gradient components:
const auto gradient_diff = var1().getElemValue(&_face_info -> elem())-var2().getElemValue(_face_info -> neighborPtr())

But that is clumsy - it would be better to directly evaluate the adCoupledGradient at the two elements. There should be a simple way of  doing that, I think.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Are you trying to build your own gradient?
We have the following ways to get you gradients here:
- var1().grad(face_arg) and var2().grad(face_arg) for face gradients
- var1().grad(elem_arg) and var2().grad(neighbor_arg) for element gradients

elem or face_arg are built from elem/neighborFromFace() routine and singleSidedFaceArg routine

Guillaume

**Reply by heinono1** [reply]
Hello Guillaume,

Thanks.! Yes, using 
ADRealVectorValue gradient_diff = var1().adGradSln(&_face_info->elem())-var2().adGradSln(_face_info->neighborPtr)));
seems to work to generate a difference in gradients across the interface. At least the compiler does not puke.

I also need to generate differences of nonlinear variables, say mag_x, across the interface. I managed (I think) to generate differences between elemental values using
params.addRequiredParam<MooseFunctorName>("mag_x",  );
..
_mag_x(getFunctor<ADReal>("mag_x")),
...
const auto mag_x_diff = _mag_x(elemFromFace()) - _mag_x(neighborFromFace());

that is, I passed the nonlinear variable as an adFunctor so I could access the elemFromFace routine. I could not use the adSln(&_face_info->elem()) for the functor or if I pass the nonlinear variable as adCoupledValue. I am not sure if it makes sense to pass the nonlinear variable mag_x as a functor. And alternative, I guess, would be to build up the interface residual by running twice with var1 and var2 being the different nonlinear variables in each call, but that seems a little clumsy.

**Reply by GiudGiud** [reply]
Retrieving it as a functor allows you to use, in addition to variables:
- functions
- functor material property
That s not useful to everyone. Many will still want to retrieve a pointer to the variable.

The flux is supposed to be the same on both sides of the interface right? Like there is conservation of whatever quantity is going through the face?

Note that coupledValue() does not allow to use the functor routines. This is for retrieving variable values the traditional way only

**Reply by heinono1** [reply]
OK, so I can retrieve the nonlinear variable as a functor - cool! No, the flux is not conserved across the interface - the gradient of the potential has a discontinuity across the interface with sources related to the nonlinear variable mag_x.

**Reply by GiudGiud** [reply]
yes

Ah i was worried about that. We dont currently allow that in FVIKs
```
void
FVInterfaceKernel::computeResidual(const FaceInfo & fi)
{
  setupData(fi);

  const auto var_elem_num = _elem_is_one ? _var1.number() : _var2.number();
  const auto var_neigh_num = _elem_is_one ? _var2.number() : _var1.number();

  const auto r = MetaPhysicL::raw_value(fi.faceArea() * fi.faceCoord() * computeQpResidual());

  processResidual(r, var_elem_num, false);
  processResidual(-r, var_neigh_num, true);
}
```
as you can see the residual is added to both sides with just a sign change. 


I actually worked on adding this for a discontinuous pressure term in Navier Stokes (for a flow diode), but it was not merged as the end product was not 100% functional.
The discontinuous flux term, with the capability to have different fluxes on both sides of an interface, was working.

I m going to tag @lindsayad for awareness and comment (and to make sure I m not delusional)

**Reply by heinono1** [reply]
Oh, I see. So I would need to be able to add two different pieces, one from each side of the interface. I can arrange so that computeQpResidual in my interface kernel calculates and returns the contributions from both sides of the interface. Would there be an easy way for me to modify my FVInterfaceKernel so the contribution I have calculated just get multiplied by the appropriate face area (without adding equal and opposite contributions from each side)?

**Reply by GiudGiud** [reply]
you ll need to overwrite ComputeResidual to make that change. computeQpResidual is only called once right now, so even if you have the right logic for distinguishing, it wont contribute a different amount on both sides

It's not where we want to be in the future but it should be doable now

**Reply by heinono1** [reply]
I'll give it a shot....

Thanks again for your help!

**Reply by lindsayad** [reply]
we could make a virtual `computeNeighborQpResidual` whose base implementation just returns the negative of `computeQpResidual`

**Reply by lindsayad** [reply]
Although that is kind of wasteful ... in most cases it would perform the AD `computeQpResidual` computation twice

**Reply by lindsayad** [reply]
I've created issue #22610 for this
